src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/PosixSignalRegistration.Unix.cs

```diff
@@ -0,0 +1,144 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Threading;
+
+namespace System.Runtime.InteropServices
+{
+    public sealed class PosixSignalRegistration : IDisposable
+    {
+        private readonly Action<PosixSignalContext> _handler;
+        private readonly PosixSignal _signal;
+        private bool _registered;
+        private readonly object _gate = new object();
+
+        private PosixSignalRegistration(PosixSignal signal, Action<PosixSignalContext> handler)
+        {
+            _signal = signal;
+            _handler = handler;
+        }
+
+        public static PosixSignalRegistration Create(PosixSignal signal, Action<PosixSignalContext> handler)
+        {
+            if (handler == null)
+            {
+                throw new ArgumentNullException(nameof(handler));
+            }
+            if (signal > PosixSignal.SIGHUP || signal < PosixSignal.SIGCHLD)
+            {
+                throw new IndexOutOfRangeException();
+            }
+            PosixSignalRegistration registration = new PosixSignalRegistration(signal, handler);
+            registration.RegisterFor(signal);
+            return registration;
+        }
+
+        private unsafe void RegisterFor(PosixSignal signal)
+        {
+            if (!s_initialized)
+            {
+                if (!Interop.Sys.InitializeTerminalAndSignalHandling())
+                {
+                    throw new Win32Exception();
+                }
+                s_initialized = true;
+            }
+            lock (s_registrations)
+            {
+                if (!s_registrations.TryGetValue(signal, out List<PosixSignalRegistration>? signalRegistrations))
+                {
+                    if (!Interop.Sys.RegisterForPosixSignal(signal, &OnPosixSignal))
+                    {
+                        throw new Win32Exception();
+                    }
+                    signalRegistrations = new List<PosixSignalRegistration>();
+                    s_registrations.Add(signal, signalRegistrations);
+                }
+                signalRegistrations.Add(this);
+            }
+            _registered = true;
+        }
+
+        private void Handle(PosixSignalContext context)
+        {
+            lock (_gate)
+            {
+                if (_registered)
+                {
+                    _handler(context);
+                }
+            }
+        }
+
+        [UnmanagedCallersOnly]
+        private static int OnPosixSignal(PosixSignal signal)
+        {
+            lock (s_registrations)
+            {
+                if (s_registrations.TryGetValue(signal, out List<PosixSignalRegistration>? signalRegistrations))
+                {
+                    if (signalRegistrations.Count != 0)
+                    {
+                        PosixSignalRegistration[] registrations = signalRegistrations.ToArray();
```

Comment:
Is this to snapshot the current list of handlers to invoke?

Reply:
Yes.